### PR TITLE
Feature/notification e learning api

### LINKED ISSUE #38: [API Improvement] : Implement Email Notification System for Course Reviews and Responses in E-Learning API
### Quick Description 🛠️

**TL;DR**: I propose adding an email notification system to inform course owners when a new review is posted, and to notify review owners when the course owner responds. This will improve engagement and communication.

---

### **Overview**
Currently, the project has an API endpoint for managing course reviews, but there is no notification system to inform course owners when a new review is posted, or to inform review owners when a course owner responds to their review. This lack of notification might result in missed feedback opportunities and poor communication between the course owners and users.

### **Proposed Solution**
I would like to implement an **email notification system** that alerts the appropriate users when:
- **A new review** is posted on their course.
- **A response** to a review is made by the course owner, notifying the review owner.

This feature will improve engagement and communication between users and course owners.

### **Key Objectives**
1. **Notify Course Owner**: 
   - When a review is submitted for one of their courses, send an email notification.
   
2. **Notify Review Owner**: 
   - When the course owner replies to a review, send an email notification to the review owner.

### **Detailed Plan**
1. **Backend Logic for Notifications**:
    - Modify the existing **review submission endpoint** to trigger an email notification when a review is created.
    - Modify the **reply-to-review endpoint** to trigger an email notification when the course owner replies.

2. **Mail Service Setup**:
    - Use **Nodemailer** (or an equivalent email service) to send email notifications.
    - Set up email templates for both notifications (new review and review response) with dynamic content such as course title, review, and response.

3. **Email Templates**:
    - Create a user-friendly email template for notifications.
    - Include information such as:
        - For course owners: **"You have received a new review on [Course Name]"**.
        - For review owners: **"The course owner has responded to your review on [Course Name]"**.

4. **Testing**:
    - Test the entire flow, ensuring notifications are sent on review submission and response.
    - Handle edge cases such as invalid email addresses or notification opt-outs!

### **Request for Feedback**
Before I begin implementing this feature, I would appreciate feedback on:
- Any specific requirements for email notifications (e.g., throttling, opt-in/out preferences).

Request: To assign me this issue, if you approve it. Along with Gssoc ext and Hacktoberfest2024 with Level tags!

### Select the open-source program you're contributing to

Other

### Adherence to Code of Conduct

- [X] I adhere to the [CONTRIBUTING GUIDELINES](../blob/master/CONTRIBUTING.md) of this project.

### PR DESCRIPTION
**Pull request by BhairabMahanta**
Implement Email Notification system for course reviews!

### Closes #38 

# 🗒️ Checklist:

- [1] I have mentioned the issue number in my Pull Request. (**done**)
- [1]  I have gone through `rules of when PR will get merged.`  (**done**)
- [1] I have commented my code, particularly in hard-to-understand areas (**done**)
- [1] I have created a helpful and easy to understand `README.md` (**done**)
- [1] I have followed proper file structure. (**done**)

- [1] I have gone through creating a CHANGES.md file for further explanation and clarity of understanding the PR, which can later be deleted. (**done**)

 Thank you for contributing!❤️ 

---
Some screenshots of working changes
image
image
image

### CODE REVIEW

**Comment by Puskar-Roy** [comment]
Hello @BhairabMahanta, thank you for your pull request! We will review it shortly. 😇❣️
Your contributions help us grow and improve our project. Please make sure you have read and followed our [CONTRIBUTING GUIDELINES](../blob/master/CONTRIBUTING.md).
Your efforts are appreciated, and we value your participation in our community. Don't forget to ⭐ our repository!
Happy coding!✨

**Comment by BhairabMahanta** [comment]
Lastly, I'd Like to thank @Puskar-Roy for allowing a long time to get used to postgres as I had been struggling to get pgadmin4 running for quite a while because of an internal laptop issue. Thanks for understanding and also giving a good opportunity to learn new things.

File: packages/APIs/E-learning_Platform_API/src/controllers/authController.ts
Comment (BhairabMahanta) [inline packages/APIs/E-learning_Platform_API/src/controllers/authController.ts:1]: There hasnt been any changes, i actually expected export interface AuthenticatedRequest extends Request {
  id: string;
}
the id to be required in login, but it was never really required!

File: packages/APIs/E-learning_Platform_API/README.md
Comment (BhairabMahanta) [inline packages/APIs/E-learning_Platform_API/README.md:1]: Only changed the lines 111-115 
others may have been affected by prettier.

File: packages/APIs/E-learning_Platform_API/CHANGES.md
Comment (BhairabMahanta) [inline packages/APIs/E-learning_Platform_API/CHANGES.md:1]: You can delete this once reviewed. Thank you.

**Review by Puskar-Roy — COMMENTED** [review]
Looks great

**Review by Puskar-Roy — APPROVED** [review]
Ok

**Review by Puskar-Roy — APPROVED** [review]
Ok

**Comment by Puskar-Roy** [comment]
@BhairabMahanta can u please raise this pr in ```dev``` branch

**Comment by Puskar-Roy** [comment]
@BhairabMahanta can u please raise this pr in ```dev``` branch

**Comment by BhairabMahanta** [comment]
> Owner

I did it!

**Comment by Puskar-Roy** [comment]
Hey @BhairabMahanta! We're excited to let you know that your pull request has been merged into the APIverse🚀🎁
Your commitment, effort, and expertise have greatly influenced our project, and we're incredibly thankful for your contributions.😇
Your code is now part of our expanding ecosystem, assisting developers worldwide in creating exceptional APIs.❣️🔥
Happy Coding! ✨ We look forward to seeing you again soon!❣️

**Comment by BhairabMahanta** [comment]
> 
> Hey @BhairabMahanta! We're excited to let you know that your pull request has been merged into the APIverse🚀🎁
> Your commitment, effort, and expertise have greatly influenced our project, and we're incredibly thankful for your contributions.😇
> Your code is now part of our expanding ecosystem, assisting developers worldwide in creating exceptional APIs.❣️🔥
> Happy Coding! ✨ We look forward to seeing you again soon!❣️
> 
> 
> 

Thanks a lot, I'll see if i can contribute more to this awesome project. 🫶🏻

**Pull request merged**